Downgrade to golang1.23 and Added validation function

Vincent056 opened the pull request:
This PR downgrades the golang to 1.23 and Add rule validation functionality and CEL expression validation

- Introduced `ValidateRule` method in the Scanner to validate rules without execution.
- Added `ValidateCELExpression` method for validating CEL expressions with given inputs.
- Created a new `validation.go` file containing the `RuleValidator` struct and related validation logic, including error categorization and detailed issue reporting for CEL expressions.
- Implemented various validation error types and structured results for validation outcomes.

Vincent056 committed:
Downgrade to golang1.23

Vincent056 committed:
Add rule validation functionality and CEL expression validation

- Introduced `ValidateRule` method in the Scanner to validate rules without execution.
- Added `ValidateCELExpression` method for validating CEL expressions with given inputs.
- Created a new `validation.go` file containing the `RuleValidator` struct and related validation logic, including error categorization and detailed issue reporting for CEL expressions.
- Implemented various validation error types and structured results for validation outcomes.

rhmdnd commented on pkg/scanner/validation_test.go:
```diff
@@ -0,0 +1,543 @@
+			wantError: false,
+		},
+		{
+			name:       "expression with undeclared variable",
```
One thing I like to do with test names is to include the outcome as part of the name. Here I might use:

```go
name:       "expression with undeclared variable fails to compile",
```

Which reinforces the behavior we're looking for in the test name - without having to jump to line 223 to see we're looking for an error. I find it a subtle thing that makes it easier to understand what the tests should do when you're running them.

rhmdnd commented on pkg/scanner/validation_test.go:
```diff
@@ -0,0 +1,543 @@
+			errorMsg:  "UNDECLARED_REFERENCE",
+		},
+		{
+			name:       "syntax error in expression",
```
Similar comment here - "syntax error in expression fails to compile"

rhmdnd commented on pkg/scanner/validation_test.go:
```diff
@@ -0,0 +1,543 @@
+			errorMsg:  "SYNTAX_ERROR",
+		},
+		{
+			name: "complex valid expression",
```
In particular, this is complex because it's using multiple membership checks (`has()`) inside a containment check (`all())`). We can elaborate on that in the test name to make it more specific - since `complex` can mean different things to different readers.

rhmdnd commented on .envrc:
```diff
@@ -0,0 +1,5 @@
+# Set GOTOOLCHAIN to local to prevent automatic Go version updates
```
This could be different for different people's environments, right? Do we want to keep this in source?

Vincent056 replied:
I think this got slipped through, let me remove it

rhmdnd commented on pkg/scanner/scanner.go:
```diff
@@ -102,12 +102,62 @@ func NewScanner(resourceFetcher ResourceFetcher, logger Logger) *Scanner {
+
+// ValidateAllRules validates all rules in a ScanConfig without executing them
+// Returns a map of rule ID to ValidationResult for detailed analysis
+func (s *Scanner) ValidateAllRules(config ScanConfig) map[string]ValidationResult {
```
I imagine the idea here is to call this from the ComplianceScan controller to validate the all rules compile before kicking off the actual scan?

Vincent056 committed:
Refactor CEL expression validation and remove .envrc

- Deleted the .envrc file as it is no longer needed.
- Renamed `validateSimpleExpression` to `validateExpression` for clarity.
- Updated validation methods to use `ValidateCELExpression` instead of `ValidateCELExpressionSimple`.
- Adjusted test cases to reflect changes in validation logic and improve error handling for CEL expressions.

rhmdnd commented on pkg/scanner/validation_test.go:
```diff
@@ -47,62 +47,60 @@ func (m *MockLogger) Error(msg string, args ...interface{}) {
+				t.Errorf("Expected validation errors, but got none")
+			}
+			if tt.expectValidationError && tt.expectedErrorType != "" && len(issues) > 0 && issues[0].Type != tt.expectedErrorType {
+				t.Errorf("Expected error type %s, got %s", tt.expectedErrorType, issues[0].Type)
```
Will all items have the same expectedErrorType?

rhmdnd reviewed (approved):
Thanks for cleaning up the tests and validation logic.

I have one more minor recommendation to reduce some of the validation logic by reusing it in a shared function, but we can implement that in a follow up commit.
